### PR TITLE
Add support to updater for updating native DEB/RPM installs with our official packages.

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

* Updated the script to actually be POSIX compliant shell script (this is an ongoing project to improve our portability).
* Restructured the install type detection to leverage the `.install-type` file if present.
* Added updater code to handle installs done with our official DEB/RPM binary packages.

##### Component Name

area/packaging

##### Test Plan

Testing should be as simple as taking the `netdata-updater.sh` script from this PR and running it on a system with an existing install with the option `--no-updater-self-update`.

### CODE REVIEW

**Comment by vkalintiris** [comment]
> Updated the script to actually be POSIX compliant shell script (this is an ongoing project to improve our portability).

Please, create a separate PR for this as it entirely unrelated to the main goal of this PR.

**Comment by Ferroin** [comment]
> > Updated the script to actually be POSIX compliant shell script (this is an ongoing project to improve our portability).
> 
> Please, create a separate PR for this as it entirely unrelated to the main goal of this PR.

It’s not _quite_ unrelated.

We actually need the updater script to be POSIX compliant for it to be guaranteed to work for native package installs. The specific issue is Debian systems, where it is possible to have a working system without bash.

**Comment by vkalintiris** [comment]
> > > Updated the script to actually be POSIX compliant shell script (this is an ongoing project to improve our portability).
> > 
> > 
> > Please, create a separate PR for this as it entirely unrelated to the main goal of this PR.
> 
> It’s not _quite_ unrelated.
> 
> We actually need the updater script to be POSIX compliant for it to be guaranteed to work for native package installs. The specific issue is Debian systems, where it is possible to have a working system without bash.

We can discuss this in a separate PR. Going full-POSIX is not a good idea (if possible at all to do without any errors creeping in other places). Changes that hurt readability and maintainability have to be justified by having at least a counter-example to a shell implementation that does not work with what we currently have. bash, dash, zsh (and even _old_ version of them) support most of the non-POSIX stuff we are using.

File: packaging/installer/netdata-updater.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/sh
```
Comment (thiagoftsm) [inline packaging/installer/netdata-updater.sh:1]: Please, change for `/bin/bash`, because `sh` is not symbolic link for `/bin/bash` on all platforms.

Comment (Ferroin) [reply packaging/installer/netdata-updater.sh:1]: The use of `/bin/sh` is intentional, see the above discussion with Vasilis.

**Comment by Ferroin** [comment]
> We can discuss this in a separate PR. Going full-POSIX is not a good idea (if possible at all to do without any errors creeping in other places). Changes that hurt readability and maintainability have to be justified by having at least a counter-example to a shell implementation that does not work with what we currently have. bash, dash, zsh (and even old version of them) support most of the non-POSIX stuff we are using.

We have to pick a specific _standard_ shell to code to if we want linting to work correctly. The only sane options are `bash` (which is _not_ portable, trivially demonstrated by Alpine (which does not include `bash` by defualt), Debian (which can be built without it), and FreeBSD and macOS (both of which require it to be explicitly installed)), and POSIX `sh`, which does work everywhere.

The only ‘problematic’ things we’re dealing with in any of our scripts are actually array usage and use of `$RANDOM` (which we only use here, and only in one place, though I do agree that _this_ specific case is less readable). More specifically:
* None of our scripts actually rely on the `=` versus `==` semantics in the `test` builtin in `bash`.
* Regex matching in conditionals has a trivial and well proven alternative in the form of `grep`.
* Variable scoping issues are properly caught and reported by shellcheck (so we don’t need `local`).
* `source` is just an alias for `.` with how we’re using it.

In reality, we don’t actually _need_ arrays in most of the places we’re using them (the only major exceptions are `packaging/installer/install-required-packages.sh`, which Igor and I have been talking about splitting into per-platform variants so we can avoid the need for associative arrays, and `health/notifications/alarm-notify.sh.in`, which we unfortunately need to keep them for for compatibility reasons) and gain little to no benefit from doing so. I’d argue most of our usage of them is actually _less_ readable than the alternatives, and in some cases, we’re even making things more complicated than they should be (see the change around line 330 in that first commit for an example).

**Comment by vkalintiris** [comment]
> > We can discuss this in a separate PR. Going full-POSIX is not a good idea (if possible at all to do without any errors creeping in other places). Changes that hurt readability and maintainability have to be justified by having at least a counter-example to a shell implementation that does not work with what we currently have. bash, dash, zsh (and even old version of them) support most of the non-POSIX stuff we are using.
> 
> We have to pick a specific _standard_ shell to code to if we want linting to work correctly. The only sane options are `bash` (which is _not_ portable, trivially demonstrated by Alpine (which does not include `bash` by defualt), Debian (which can be built without it), and FreeBSD and macOS (both of which require it to be explicitly installed)), and POSIX `sh`, which does work everywhere.
> 
> The only ‘problematic’ things we’re dealing with in any of our scripts are actually array usage and use of `$RANDOM` (which we only use here, and only in one place, though I do agree that _this_ specific case is less readable). More specifically:
> 
> * None of our scripts actually rely on the `=` versus `==` semantics in the `test` builtin in `bash`.
> * Regex matching in conditionals has a trivial and well proven alternative in the form of `grep`.
> * Variable scoping issues are properly caught and reported by shellcheck (so we don’t need `local`).
> * `source` is just an alias for `.` with how we’re using it.
> 
> In reality, we don’t actually _need_ arrays in most of the places we’re using them (the only major exceptions are `packaging/installer/install-required-packages.sh`, which Igor and I have been talking about splitting into per-platform variants so we can avoid the need for associative arrays, and `health/notifications/alarm-notify.sh.in`, which we unfortunately need to keep them for for compatibility reasons) and gain little to no benefit from doing so. I’d argue most of our usage of them is actually _less_ readable than the alternatives, and in some cases, we’re even making things more complicated than they should be (see the change around line 330 in that first commit for an example).

@Ferroin I agree 99% with what you explained. My primary motivation is to have this discussion in a separate PR. Having big PRs that try to achieve multiple goals at the same time makes reviews difficult to follow and longer to merge.

**Comment by vkalintiris** [comment]
@Ferroin the PR is marked as a draft. Can you rebase it?

**Comment by Ferroin** [comment]
> @Ferroin the PR is marked as a draft. Can you rebase it?

It’s still marked as a draft because #11755 has (intentionally) not been merged yet. There’s not much point in rebasing it either until that gets merged.

File: packaging/installer/netdata-updater.sh [hunk 2]
```diff
@@ -396,17 +554,26 @@ done
...
 fi
 
 # shellcheck source=/dev/null
-source "${ENVIRONMENT_FILE}" || exit 1
+. "${ENVIRONMENT_FILE}" || exit 1
+
+if [ -f "$(dirname "${ENVIRONMENT_FILE}")/.install-type" ]; then
+  # shellcheck source=/dev/null
+  . "$(dirname "${ENVIRONMENT_FILE}")/.install-type" || exit 1
+fi
```
Comment (vkalintiris) [inline packaging/installer/netdata-updater.sh:570]: We could have an `INSTALL_TYPE_FILE` variable right above `safe_sha256sum`'s definition and just source it.

File: packaging/installer/netdata-updater.sh [hunk 3]
```diff
@@ -371,6 +369,166 @@ update() {
...
+
+  sysarch="$(uname -m)"
+
+  download "${NETDATA_TARBALL_CHECKSUM_URL}" "${ndtmpdir}/sha256sum.txt"
+  download "${NETDATA_TARBALL_URL}" "${ndtmpdir}/netdata-${sysarch}-latest.gz.run"
+  if ! grep netdata-latest.gz.run "${ndtmpdir}/sha256sum.txt" | safe_sha256sum -c - > /dev/null 2>&1; then
+    fatal "Static binary checksum validation failed. Stopping netdata installation and leaving binary in ${ndtmpdir}\nUsually this is a result of an older copy of the file being cached somewhere and can be resolved by simply retrying in an hour."
+  fi
+
+  if [ -e /opt/netdata/etc/netdata/.install-type ] ; then
```
Comment (vkalintiris) [inline packaging/installer/netdata-updater.sh:387]: `s/-e/-r/`

**Comment by Ferroin** [comment]
Rebased to pick up assorted changes.

**Comment by Ferroin** [comment]
Oops, wrong button....

File: packaging/installer/netdata-updater.sh [hunk 4]
```diff
@@ -369,6 +369,167 @@ update() {
...
+  fi
+
+  # shellcheck disable=SC1090
+  . "${os_release_file}"
+
+  DISTRO="${ID}"
+
+  supported_compat_names="debian ubuntu centos fedora opensuse"
+
+  if str_in_list "${DISTRO}" "${supported_compat_names}"; then
```
Comment (vkalintiris) [inline packaging/installer/netdata-updater.sh:426]: Where is `str_in_list` defined? Same for `INTERACTIVE` below. If we expect these to come from env then list them explicitly in the 1st comment of this file.

File: packaging/installer/netdata-updater.sh [hunk 5]
```diff
@@ -369,6 +369,167 @@ update() {
...
+    # shellcheck disable=SC2086
+    env ${env} ${pm_cmd} ${repo_subcmd} ${repo_update_opts} || fatal "Failed to update repository metadata."
+  fi
+
+  if ${pkg_installed_check} netdata-repo-edge > /dev/null 2>&1; then
+    # shellcheck disable=SC2086
+    env ${env} ${pm_cmd} ${upgrade_cmd} ${pkg_install_opts} netdata-repo-edge || fatal "Failed to update Netdata repository config."
+    # shellcheck disable=SC2086
+    env ${env} ${pm_cmd} ${repo_subcmd} ${repo_update_opts} || fatal "Failed to update repository metadata."
+  fi
```
Comment (vkalintiris) [inline packaging/installer/netdata-updater.sh:527]: Can we use a loop or a separate function for this?

File: packaging/installer/netdata-updater.sh (hunk 4 above)
Comment (Ferroin) [reply packaging/installer/netdata-updater.sh:426]: Both are missing definitions that I forgot to deal with when copying some of this code over from kickstart-ng.sh.

File: packaging/installer/netdata-updater.sh [hunk 6]
```diff
@@ -371,6 +369,166 @@ update() {
   return 0
 }
 
+update_static() {
+  ndtmpdir="$(create_tmp_directory)"
+  PREVDIR="$(pwd)"
+
+  echo >&2 "Entering ${ndtmpdir}"
```
[the remark this answers is not in the document]
Comment (ilyam8) [reply packaging/installer/netdata-updater.sh]: Done in https://github.com/netdata/netdata/pull/11879, should we merge #11879 first? @Ferroin

[the remark this answers is not in the document]
Comment (vkalintiris) [reply packaging/installer/netdata-updater.sh]: @ilyam8 fwiw, my comment was for the changes of this PR.

[the remark this answers is not in the document]
Comment (Ferroin) [reply packaging/installer/netdata-updater.sh]: @ilyam8 Due to how things would end up shuffled around here, I’m not sure if it matters, but I did just doe a pass over the whole file in the most recent commit to fix things so we’re properly using info/warning/fatal where appropriate.

**Comment by Ferroin** [comment]
Rebased to pick up latest updater changes.

**Review by iigorkarpov — APPROVED** [review]

**Comment by Ferroin** [comment]
Rebased to pick up recent updater code changes.

**Review by iigorkarpov — DISMISSED** [review]

**Comment by thiagoftsm** [comment]
@Ferroin I would like to confirm something with you, because I think we have a missing something in your PR. I tested on CentOS 8.x, and I observed that the PR worked as I expected, but when I tested on Debian 10.x and Slackware current I received the following error:

VirtualBox_Slackware_current_05_01_2022_00_33_01

I was expecting a message like next image that I receive when I use master branch:

VirtualBox_Slackware_current_05_01_2022_00_32_32

Best regards!

**Comment by Ferroin** [comment]
@thiagoftsm  For testing this, you need to pass the option `--no-updater-self-update` when invoking the updater, otherwise it will try to use the copy from the `master` branch here on GitHub instead of the code we’re trying to test.

That said, that should not be causing the issues you mention. Can you share the `.environment` files from the installs you were seeing issues on (and confirm that you get correct output from the updater script in the `master` branch of the GitHub repository)?

**Comment by thiagoftsm** [comment]
> @thiagoftsm For testing this, you need to pass the option `--no-updater-self-update` when invoking the updater, otherwise it will try to use the copy from the `master` branch here on GitHub instead of the code we’re trying to test.

I tested with the option and without it.

> That said, that should not be causing the issues you mention. Can you share the `.environment` files from the installs you were seeing issues on (and confirm that you get correct output from the updater script in the `master` branch of the GitHub repository)?

Of course, the Debian output is this:

VirtualBox_Debian10_05_01_2022_13_33_20

While the Slackware current is this:

VirtualBox_Slackware_current_05_01_2022_13_17_45

Let me know if you need any other information.

**Comment by Ferroin** [comment]
@thiagoftsm Can you confirm if either system has a `.install-type` file alongside the `.environment` file, and if so what the contents are?

**Comment by Ferroin** [comment]
@thiagoftsm  Correction, I just noticed an issue in the install type detection, and have now pushed a fix for that as well as updating the error mesages to be better for unsupported install types. Can you please re-test with the latest code (it probably will still fail, but the error message will help diagnose what, exactly, is going wrong here).

**Comment by thiagoftsm** [comment]
> @thiagoftsm Can you confirm if either system has a `.install-type` file alongside the `.environment` file, and if so what the contents are?
Both have the same content:

VirtualBox_Slackware_current_05_01_2022_14_16_13

I will test now the fix your brought.

**Comment by Ferroin** [comment]
> > @thiagoftsm Can you confirm if either system has a `.install-type` file alongside the `.environment` file, and if so what the contents are?
> > Both have the same content:
> 
> VirtualBox_Slackware_current_05_01_2022_14_16_13
> 
> I will test now the fix your brought.

OK, if both show `custom` for the install type, then they were likely installed manually instead of using the kickstart script. The code as it is now does not allow updating such installs using the updater script, but there is no technical reason why such an install installed using the `netdata-installer.sh` script shouldn’t work correctly, so I’ll look at getting that fixed. With the latest code, you should see an error message on these systems reading: `This script does not support updating custom installations.`.

**Comment by Ferroin** [comment]
@thiagoftsm I’ve further updated the install type detection. The latest code _should_ work now for both of those systems you were seeing issues with.

**Comment by thiagoftsm** [comment]
> @thiagoftsm I’ve further updated the install type detection. The latest code _should_ work now for both of those systems you were seeing issues with.

I have different messages on Debian

VirtualBox_Debian10_05_01_2022_15_03_19

and Slackware:

VirtualBox_Slackware_current_05_01_2022_14_39_09

in both cases it was possible to update.  I also retest on CentOS 8.5 and the update worked.

**Review by thiagoftsm — APPROVED** [review]
As described [here](https://github.com/netdata/netdata/pull/11753#issuecomment-1005798264), the tests worked, so LGTM!

**Comment by Ferroin** [comment]
> > @thiagoftsm I’ve further updated the install type detection. The latest code _should_ work now for both of those systems you were seeing issues with.
> 
> I have different messages on Debian
> 
> VirtualBox_Debian10_05_01_2022_15_03_19
> 
> and Slackware:
> 
> VirtualBox_Slackware_current_05_01_2022_14_39_09
> 
> in both cases it was possible to update. I also retest on CentOS 8.5 and the update worked.

You ran the uninstaller instead of the updater on Slackware, hence the different messages.

**Comment by maneamarius** [comment]
I have some questions, do we test netdata-updater anywhere in our PR tests? (I can't seem to see that)
If not, how can we make sure this script will work properly on ALL supported platforms?
Also, did you try running the script with "set -e"? (we should aim having that by default going forward). This will allow us catch any silent errors.

**Comment by Ferroin** [comment]
> Do we test netdata-updater anywhere in our PR tests? (I can't seem to see that) If not, how can we make sure this script will work properly on ALL supported platforms?

Yes, but not rigorously, and it does not cover the particular case involved here. The relevant checks are in the Build workflow, and are named ‘Test Generated Distfile and Updater Code’. Right now they only test running the updater code in the PR against an existing install using the distfiles generated from the PR (in 99% of cases this type of check is what we want, it’s just not as comprehensive as we want).

We will need to expand those long-term (both for this new code, and to cover the static build cases), but the complexity involved is arguably a bit much for this PR (it will need a reorganization of code in the Packages workflow). I’m planning on doing another PR to cover those additions at some point in the relatively near future.

> Also, did you try running the script with "set -e"? (we should aim having that by default going forward). This will allow us catch any silent errors.

`set -e` breaks this particular script right now because we’re relying in a couple of places on expected errors being silently ignored. I agree that long-term we should have a goal of using it in all of our scripts, but the required changes for the updater code are to work with this are kind of invasive, and this type of fix is orthogonal to what this PR is doing, so it should be in it’s own PR.

**Review by maneamarius — APPROVED** [review]

**Pull request merged**